src/main/java/io/cdap/plugin/splunk/common/client/SplunkSearchIterator.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.splunk.common.client;
+
+import com.github.rholder.retry.RetryException;
+import com.github.rholder.retry.Retryer;
+import com.google.common.annotations.VisibleForTesting;
+import com.splunk.Event;
+import com.splunk.Job;
+import com.splunk.ResultsReader;
+import com.splunk.ResultsReaderCsv;
+import com.splunk.ResultsReaderJson;
+import com.splunk.ResultsReaderXml;
+import com.splunk.Service;
+import io.cdap.plugin.splunk.common.exception.ConnectionTimeoutException;
+import io.cdap.plugin.splunk.common.exception.JobResultsException;
+import io.cdap.plugin.splunk.common.util.SearchHelper;
+import io.cdap.plugin.splunk.source.batch.SplunkBatchSourceConfig;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.regex.Pattern;
+
+/**
+ * Iterator for Splunk Search.
+ */
+public class SplunkSearchIterator implements Iterator<Map<String, String>>, Closeable {
+
+  private static final Pattern RESTRICTED_PATTERN = Pattern.compile("[^a-zA-Z0-9_]");
+
+  private final Service splunkService;
+  private final SplunkBatchSourceConfig config;
+  private final String searchId;
+  private final Long offset;
+  private final Long count;
+
+  private ResultsReader resultsReader;
+  private Iterator<Event> iterator;
+
+  public SplunkSearchIterator(Service splunkService, SplunkBatchSourceConfig config,
+                              String searchId, Long offset, Long count) {
+    this.splunkService = splunkService;
+    this.config = config;
+    this.searchId = searchId;
+    this.offset = offset;
+    this.count = count;
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (resultsReader == null || iterator == null) {
+      try {
+        InputStream stream = getStreamResults(config, searchId, offset, count);
+        resultsReader = getResultsReader(stream, config.getOutputFormat());
+        iterator = resultsReader.iterator();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
```

Comment:
I think we should throw IllegalStateException at least

Reply:
Done.